### PR TITLE
[RFC] TypeScript Refactor

### PR DESCRIPTION
**Pull request by blainekasten**
## TypeScript Refactor

[Rendered RFC](https://github.com/gatsbyjs/gatsby/blob/8f9e722e2fb073b83b006408ba3e061d6a379ca2/rfcs/typescript-refactor.md)

### CODE REVIEW

**Comment by pvdz** [comment]
(I don't know whether learning is supposed to take a look at these...?)

File: rfcs/typescript-refactor.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+Flow. Unfortunately, it is less mature, has a less robust ecosystem, and will require us to manually type 3rd party libraries. The main downside to Flow is the non-strictness it has. TypeScript can fail builds if it is not typed soundly, and this is a huge benefit actually. Flow not having that allows you to run down rabbit holes of unsound code until you finlly run aflow and are told of your problems.
+
+Other alternatives would be a different language like Reason or Rust + WASM.
+
+# Adoption strategy
+
+Here are the steps to setup a package to use TS:
+
+1. Add a tsconfig.json to the package with an `"extends": "../../tsconfig.json"` entry.
+2. Begin changing files from .js to .ts and adding types!
```
Comment (tesseralis) [inline rfcs/typescript-refactor.md]: Is there a concrete idea of what direction the integration should go? Should we start with utility files with little dependencies first and move out? Or start with the most public facing files like APIs and fill in the gaps? Or is it just a free-for-all?

**Comment by dominicfallows** [comment]
+1 it would be great to start having strongly typed code in the core.

File: rfcs/typescript-refactor.md (hunk 1 above)
Comment (blainekasten) [reply rfcs/typescript-refactor.md]: This is our approach at the start:

- If you are working on a feature or fix, any file you touch consider if you can easily switch it to TS. If so, do it.
- Else if you have downtime, or are unsure what to work on, switch one of the “harder” files to TS.

These aren't hard and fast rules. But it's basically just guidelines to say, don't drop what you're doing to work on TS. And don't ignore it either. haha. I'm going to try and wrangle community support for this as well so we don't have to do all of the heavy lifting!

File: rfcs/typescript-refactor.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+# Detailed design
+
+The following packages will be converted to TypeScript:
+
+- babel-plugin-remove-graphql-queries
+- babel-preset-gatsby
+- babel-preset-gatsby-package
+- gatsby
+- gatsby-cli
+- gatsby-codemods
```
Comment (blainekasten) [inline rfcs/typescript-refactor.md]: this one is unnecessary

**Pull request merged**